posts/simprocs-intro.md

```diff
@@ -0,0 +1,550 @@
+---
+author: 'Yaël Dillies, Paul Lezeau'
+category: 'Metaprogramming'
+date: 2025-03-29 12:33:00 UTC+00:00
+description: 'Brief introduction to simprocs and what they are useful for'
+has_math: true
+link: ''
+slug: what-is-a-simproc
+tags: 'simp, simproc, meta'
+title: What is a simproc?
+type: text
+---
+
+Lean core recently added support for custom simplification procedures, aka *simprocs*.
+This blog post aims to explain what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.
+
+<!-- TEASER_END -->
+
+The first part of this post is a purely informal description of what simprocs are and do.
+The second part is a walkthrough to writing a simple simproc in three different ways.
+
+# What is a simproc
+
+To understand what a simproc is and how it works, we will first explain how `simp` works.
+Then we will give some examples and non-examples of simprocs as well as pointers to analogous concepts in other theorem provers.
+
+## How simp works
+
+On an informal level, `simp` can be seen as recursively traversing the expression `e` to be simplified from the outside-in, each time checking whether the expression matches the left hand side `LHS` of some simp lemma of the form `LHS = RHS`, and replace `LHS` by `RHS` in the expression if it does match.
+
+To see what steps `simp` performs, one can use `set_option trace.Meta.Tactic.simp true`, which prints some of this information in the infoview.
+
+Here is an example of a series of rewrites performed by `simp` on a simple goal:
+```lean
+set_option trace.Meta.Tactic.simp true in
+example : 37 * (Nat.fib 0 + 0) = 0 := by
+  simp
+```
+which yields
+```
+[Meta.Tactic.simp.rewrite] Nat.fib_zero:1000:
+      Nat.fib 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] add_zero:1000:
+      0 + 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] mul_zero:1000:
+      37 * 0
+    ==>
+      0
+[Meta.Tactic.simp.rewrite] eq_self:1000:
+      0 = 0
+    ==>
+      True
+```
+i.e. the steps are:
+```lean
+⊢ 37 * (Nat.fib 0 + 0) = 0
+⊢ 37 * (0 + 0) = 0
+⊢ 37 * 0 = 0
+⊢ 0 = 0
+⊢ True
+```
+
+In this picture, simp lemmas are *fixed* rules to turn a *specific* left hand side into a *specific* right hand side.
+In contrast, simprocs are *flexible* rules to turn a *specific* left hand side into a right hand side *computed* from the left hand side.
+In this sense, they are *parametric simp lemmas*.
```

Comment:
To a beginner, I don't think this is a useful distinction without some more explanation. The lemma `a + 0 = a` is "parametric" in the sense that I can feed in any `a`, and the lemma still applies, indeed `a` is a parameter here! There's also a reasonable interpretation of "specific" which makes this example not a simp lemma: it's flexible in what `a` can be, I haven't specified the value of `a`.

In particular, the description below, "combinatorial explosion of lemmas"  - to a beginner - doesn't clearly distinguish between `1 + 0 = 1`, `2 + 0 = 2`, `3 + 0 = 3`, etc as a combinatorial explosion of lemmas at this point either, I can imagine someone thinking `a + 0 = a` isn't solvable without simprocs from this explanation.

Reply:
I think this is **the** point of simprocs, so I really want to keep that distinction. The blurb now refers to the "shape" of the LHS and RHS, rather than to the LHS and RHS directly. Hopefully that clears things up enough?